### PR TITLE
(0.105.4) Add basic tests for Reactant-Oceananigans correctness

### PR DESCRIPTION
**Pull request by glwagner**
This PR starts to port functionality over from [GB-25](https://github.com/PRONTOLab/GB-25) for testing the correctness of reactant-compiled kernels that perform basic Oceananigans tasks, like filling halos or computing tendencies.

This PR starts with tests for fill halo regions. 

cc @giordano @wsmoses

### CODE REVIEW

File: test/test_reactant_correctness.jl [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                                    Random.seed!(12345)
+                                    init_data = randn(Float64, size(vanilla_field)...)
+
+                                    # Set fields with the same initial data
+                                    set!(vanilla_field, init_data)
+                                    set!(reactant_field, init_data)
+
+                                    # Fill halos on both
+                                    fill_halo_regions!(vanilla_field)
+                                    @jit fill_halo_regions!(reactant_field)
```
Comment (wsmoses) [inline test/test_reactant_correctness.jl]: You should test with raise=true and also raise=false

For auto diff to work you need to set raise=true and raise first=true

Comment (glwagner) [reply test/test_reactant_correctness.jl]: @wsmoses can you take another look? looks like the fill halo regions tests pass now (is that expected?) and we are on to a LatitudeLongitudeGrid issue

**Comment by glwagner** [comment]
@wsmoses @jlk9 it looks like a few things don't work with `LatitudeLongitudeGrid` --- either `coriolis=HydrostaticSphericalCoriolis()` or `advection=WENO()`. If this jives with what you've seen I can try to generate a non-Oceananigans MWE.

**Comment by glwagner** [comment]
@jtbuch and @dkytezab are interested in this

**Comment by dkytezab** [comment]
@glwagner seems like the problem also applies to `RectilinearGrid` if you change the `x, y` boundary conditions to `Bounded`.

File: test/test_reactant_correctness.jl [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    @testset "compute_simple_Gu! correctness" begin
+        advection_schemes = (nothing, Centered(), WENO())
+        
+        # Helper to get advection name for testset
+        adv_name(::Nothing) = "nothing"
+        adv_name(a) = string(nameof(typeof(a)))
+
+        # RectilinearGrid tests
+        @testset "RectilinearGrid" begin
+            kw = (size=(4, 4, 4), halo=(3, 3, 3), extent=(1, 1, 1), topology=(Periodic, Periodic, Bounded))
```
Comment (glwagner) [inline test/test_reactant_correctness.jl]: @dkytezab your comment pertains to this right? So we should loop over topologies for both Gu and the integration test. I can change that.

Comment (dkytezab) [reply test/test_reactant_correctness.jl]: Yeah exactly

Comment (dkytezab) [reply test/test_reactant_correctness.jl]: It gets weirder... `RectilinearGrid` with `(Bounded, Periodic, Bounded)` errs but `(Periodic, Bounded, Bounded)` does not err.

Comment (glwagner) [reply test/test_reactant_correctness.jl]: that is pretty interesting. This CAN happen because there is an issue of "who owns the corners" which requires sopme care when mixing Bounded and Periodic

**Comment by dkytezab** [comment]
1. So the RectilinearGrid + LatittudeLongitudeGrid `fill_halo_regions!` tests pass if we upgrade to Julia 1.11.. though `TripolarGrid` fails due to a failure to raise. Worth investigating further
2. RectilinearGrid + LatitudeLongitudeGrid tendency computation tests are failing because of an inability to raise when using WENO **and** when the x-axis is `Bounded`.
3. HydrostaticFreeSurfaceModel with at least one Bounded x/y axis throws a dimension mismatch issue

**Comment by giordano** [comment]
Interesting.  I think that'd be ok, as long as it doesn't break other Enzyme/Reactant tests.

**Comment by giordano** [comment]
Maybe we can try the upgrade to julia v1.11 in a separate PR?  Need to change https://github.com/CliMA/Oceananigans.jl/blob/b1b07fe1572464ac8361fa6ea579545f6c5412de/.buildkite/pipeline.yml#L3 to 1.11.9

**Comment by dkytezab** [comment]
That would be perfect

**Comment by dkytezab** [comment]
I think the fix worked

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -198,9 +198,10 @@ steps:
           - "CPU"
           - "GPU"
         group:
-          - "🎭 reactant_2"
-          - "👺 enzyme"
-          - "👹 reactant_1"
+          # - "🎭 reactant_2"
```
Comment (giordano) [inline .buildkite/pipeline.yml]: Commented out tests should be restored before merge, right?

File: Project.toml [hunk 4]
```diff
@@ -5,6 +5,8 @@ authors = ["Climate Modeling Alliance and contributors"]
 
 [deps]
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+ConstructionBase = "187b0558-2788-49d3-abe0-74a17ed4e7c9"
```
Comment (giordano) [inline Project.toml]: Adding these two packages as strong deps is wrong, right?

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -198,9 +198,10 @@ steps:
...
+          # - "👺 enzyme"
+          # - "👹 reactant_1"
```
Comment (glwagner) [inline .buildkite/pipeline.yml]:
```suggestion
          - "🎭 reactant_2"
          - "👺 enzyme"
          - "👹 reactant_1"
```

File: Project.toml (hunk 4 above)
Comment (glwagner) [inline Project.toml]:
```suggestion
```

File: ext/OceananigansReactantExt/Grids/Grids.jl [hunk 6]
```diff
@@ -64,18 +64,18 @@ reactant_offset_indices(ℓ,         topo, N, H, ::Colon) = reactant_offset_indi
 reactant_offset_indices(ℓ,         topo, N, H, r::AbstractUnitRange) = r
 reactant_offset_indices(::Nothing, topo, N, H, ::AbstractUnitRange) = 1:1
 
-function Oceananigans.Grids.new_data(FT::DataType, arch::Union{ReactantState, ShardedDistributed},
+function Oceananigans.Grids.new_data(FT::DataType, arch::ShardedDistributed,
```
Comment (glwagner) [inline ext/OceananigansReactantExt/Grids/Grids.jl:67]: @romanlee note these changes here

File: test/runtests.jl [hunk 7]
```diff
@@ -279,7 +280,16 @@ CUDA.allowscalar() do
...
+    =#
 
+    # Tests for Reactant correctness (comparing vanilla vs ReactantState)
+    if group == :reactant_correctness || group == :all
+        @testset "Reactant correctness tests" begin
+            include("test_reactant_correctness.jl")
+        end
+    end
+
+    #=
```
Comment (glwagner) [inline test/runtests.jl]:
```suggestion
```

File: test/runtests.jl [hunk 8]
```diff
@@ -333,6 +342,7 @@ CUDA.allowscalar() do
     if group == :convergence
         include("test_convergence.jl")
     end
+    =#
```
Comment (glwagner) [inline test/runtests.jl]:
```suggestion
```

File: test/runtests.jl [hunk 9]
```diff
@@ -279,6 +280,14 @@ CUDA.allowscalar() do
             include("test_reactant_latitude_longitude_grid.jl")
         end
     end
+    =#
```
Comment (glwagner) [inline test/runtests.jl]:
```suggestion
   # Initialization steps
   if group == :init || group == :all
       include("test_init.jl")
@@ -279,6 +280,14 @@ CUDA.allowscalar() do
           include("test_reactant_latitude_longitude_grid.jl")
       end
   end
```

File: test/test_reactant_correctness.jl [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+                        vanilla_field = Field{LX, LY, LZ}(vanilla_grid)
+                        reactant_field = Field{LX, LY, LZ}(reactant_grid)
+
+                        Random.seed!(12345)
+                        data = randn(size(vanilla_field)...)
+                        set!(vanilla_field, data)
+                        set!(reactant_field, data)
+
+                        fill_halo_regions!(vanilla_field)
+                        @jit fill_halo_regions!(reactant_field)
```
Comment (glwagner) [inline test/test_reactant_correctness.jl]:
```suggestion
                        rfill! = @compile raise=true raise_first=true fill_halo_regions!(reactant_field)
                        rfill!(reactant_field)
```

File: test/test_reactant_correctness.jl [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+                        vanilla_field = Field{LXf, LYf, LZf}(vanilla_grid)
+                        reactant_field = Field{LXf, LYf, LZf}(reactant_grid)
+
+                        Random.seed!(12345)
+                        data = randn(size(vanilla_field)...)
+                        set!(vanilla_field, data)
+                        set!(reactant_field, data)
+
+                        fill_halo_regions!(vanilla_field)
+                        @jit fill_halo_regions!(reactant_field)
```
Comment (glwagner) [inline test/test_reactant_correctness.jl]:
```suggestion
                        rfill! = @compile raise=true raise_first=true fill_halo_regions!(reactant_field)
                        rfill!(reactant_field)
```

**Comment by dkytezab** [comment]
Currently, raising WENO with bounded-x axis on `RectilinearGrid` is broken. Same with raising WENO on `LatitudeLongitudeGrid`.

For the moment, lets flag this and compile those failing cases without raising them.

File: test/test_reactant_correctness.jl [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+                        @test compare_parent("halo", vanilla_field, reactant_field)
+                    end
+                end
+            end
+        end
+
+        # TripolarGrid tests commented out: Reactant MLIR segfaults
+        # in CanonicalizeLoopsPass when compiling Zipper boundary conditions
+        # on TripolarGrid. This crash is non-deterministic and kills the
```
Comment (wsmoses) [inline test/test_reactant_correctness.jl:101]: does this still happen? if so this is rather critical to mwe/debug?

Comment (glwagner) [reply test/test_reactant_correctness.jl:101]: i can work on this

File: Project.toml (hunk 4 above)
Comment (glwagner) [reply Project.toml]: pretty sure yeah

File: test/test_reactant_correctness.jl (hunk 12 above)
Comment (glwagner) [reply test/test_reactant_correctness.jl:101]: it is critical if we want to use tripolar for sure. i can look at it

File: test/test_reactant_correctness.jl [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+                        rfill!(reactant_field)
+
+                        @test compare_parent("halo", vanilla_field, reactant_field)
+                    end
+                end
+            end
+        end
+
+        # Test LatitudeLongitudeGrid with LX ∈ (Periodic, Bounded)
+        for LX in (Periodic, Bounded)
```
Comment (glwagner) [inline test/test_reactant_correctness.jl:70]:
```suggestion
        for LX in (Periodic,)
```

File: src/Grids/vertical_discretization.jl [hunk 14]
```diff
@@ -278,7 +278,7 @@ z_domain(grid) = domain(topology(grid, 3)(), grid.Nz, grid.z.cᵃᵃᶠ)
     Nz = size(grid, 3)
     nodes = rnodes(grid, Face(); with_halos=true)
     cpu_nodes = on_architecture(CPU(), nodes)
-    return cpu_nodes[1:Nz+1]
+    return view(cpu_nodes, 1:Nz+1)
```
Comment (glwagner) [inline src/Grids/vertical_discretization.jl]:
```suggestion
    return cpu_nodes[1:Nz+1]
```

File: ext/OceananigansReactantExt/OceananigansReactantExt.jl [hunk 15]
```diff
@@ -314,6 +314,16 @@ end
 
 const ReactantClock = Oceananigans.TimeSteppers.Clock{<:Any, <:Any, <:Reactant.TracedRNumber}
 
+# Convert Clock to a NamedTuple for kernel args.
+# Workaround for Reactant/MLIR bug where mutable structs with traced scalar fields
+# in tuples generate mismatched LLVM function signatures (extra phantom pointer).
+Oceananigans.TimeSteppers.kernel_clock(clock::Oceananigans.TimeSteppers.Clock) =
```
Comment (wsmoses) [inline ext/OceananigansReactantExt/OceananigansReactantExt.jl]: this is the wrong fix, please open an mwe against reactant with the error

Comment (glwagner) [reply ext/OceananigansReactantExt/OceananigansReactantExt.jl]: working on it

Comment (glwagner) [reply ext/OceananigansReactantExt/OceananigansReactantExt.jl]: actually this may be an excellent clue about what's going wrong

Comment (glwagner) [reply ext/OceananigansReactantExt/OceananigansReactantExt.jl]: `Clock` is a mutable and we have this adapt_structure

https://github.com/CliMA/Oceananigans.jl/blob/c0ad19bff6651f512bedbb7cd6508376b4504590/src/TimeSteppers/clock.jl#L154-L158

File: src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl [hunk 16]
```diff
@@ -185,12 +185,14 @@ end
 @kernel function compute_hydrostatic_free_surface_Gu!(Gu, grid, args)
     i, j, k = @index(Global, NTuple)
     @inbounds Gu[i, j, k] = hydrostatic_free_surface_u_velocity_tendency(i, j, k, grid, args...)
+    nothing
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl]:
```suggestion
```

File: src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl [hunk 17]
```diff
@@ -191,6 +191,7 @@ end
 @kernel function compute_hydrostatic_free_surface_Gv!(Gv, grid, args)
     i, j, k = @index(Global, NTuple)
     @inbounds Gv[i, j, k] = hydrostatic_free_surface_v_velocity_tendency(i, j, k, grid, args...)
+    nothing
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl]:
```suggestion
```

File: src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl [hunk 18]
```diff
@@ -201,4 +201,5 @@ end
 @kernel function compute_hydrostatic_free_surface_Gc!(Gc, grid, args)
     i, j, k = @index(Global, NTuple)
     @inbounds Gc[i, j, k] = hydrostatic_free_surface_tracer_tendency(i, j, k, grid, args...)
+    nothing
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_tendencies.jl]:
```suggestion
```

File: src/Models/HydrostaticFreeSurfaceModels/update_hydrostatic_free_surface_model_state.jl [hunk 19]
```diff
@@ -69,8 +69,11 @@ function update_state!(model::HydrostaticFreeSurfaceModel, grid, callbacks)
 
     # Fill only local halos for diagnostic quantities since the parameters used
     # above include regions inside the (horizontal) halos.
-    fill_halo_regions!(model.closure_fields; only_local_halos=true)
-    fill_halo_regions!(model.pressure.pHY′; only_local_halos=true)
+    # Note: we pass model.clock and fields(model) to avoid empty args tuple,
+    # which causes Reactant MLIR compilation failures (empty tuples get
+    # eliminated at call sites but not in function definitions).
+    fill_halo_regions!(model.closure_fields, model.clock, fields(model); only_local_halos=true)
+    fill_halo_regions!(model.pressure.pHY′, model.clock, fields(model); only_local_halos=true)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/update_hydrostatic_free_surface_model_state.jl]:
```suggestion
    fill_halo_regions!(model.closure_fields; only_local_halos=true)
    fill_halo_regions!(model.pressure.pHY′; only_local_halos=true)
```

File: src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_buffers.jl [hunk 20]
```diff
@@ -44,7 +44,7 @@ function complete_communication_and_compute_momentum_buffer!(model::HydrostaticF
     update_hydrostatic_pressure!(model.pressure.pHY′, arch, grid, model.buoyancy, model.tracers; parameters = surface_params)
     compute_closure_fields!(model.closure_fields, model.closure, model; parameters = κ_params)
 
-    fill_halo_regions!(model.closure_fields; only_local_halos=true)
+    fill_halo_regions!(model.closure_fields, model.clock, fields(model); only_local_halos=true)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/compute_hydrostatic_free_surface_buffers.jl]:
```suggestion
    fill_halo_regions!(model.closure_fields; only_local_halos=true)
```

File: ext/OceananigansReactantExt/Architectures.jl [hunk 21]
```diff
@@ -29,6 +29,10 @@ on_architecture(::ReactantState, a::Reactant.AnyTracedRArray) = a
 on_architecture(::CPU, a::AnyConcreteReactantArray) = Array(a)
 on_architecture(::CPU, a::SubArray{<:Any,<:Any,<:AnyConcreteReactantArray}) = Array(a)
 
+using OffsetArrays: OffsetArray
+on_architecture(::CPU, a::OffsetArray{<:Any,<:Any,<:AnyConcreteReactantArray}) =
```
Comment (glwagner) [inline ext/OceananigansReactantExt/Architectures.jl]:
```suggestion
on_architecture(::CPU, a::OffsetArray{<:Any, <:Any, <:AnyConcreteReactantArray}) =
```

**Comment by glwagner** [comment]
@wsmoses tests pass on my mac so this may be read to merge

**Comment by glwagner** [comment]
drat

File: test/test_reactant_correctness.jl [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+            # could also test these very uncommonly-used topologies:
+            #(Bounded, Periodic, Bounded),
+            #(Bounded, Bounded, Periodic),
+            #(Periodic, Bounded, Periodic),
+            #(Bounded, Periodic, Periodic),
+            #(Periodic, Periodic, Periodic),
+    )
+
+    # Note: raise=true mode is needed for autodiff but triggers non-deterministic
+    # segfaults in Reactant's CanonicalizeLoopsPass (MLIR bug). All tests below
```
Comment (wsmoses) [inline test/test_reactant_correctness.jl]: is there an issue open for this that can be linked to?

Comment (glwagner) [reply test/test_reactant_correctness.jl]: i think this is out of date bc the tests below do use raise_first=true

File: test/test_reactant_correctness.jl [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+    # use only the default raise=false mode until this is resolved upstream.
+
```
Comment (glwagner) [inline test/test_reactant_correctness.jl]:
```suggestion
```

File: test/test_reactant_correctness.jl (hunk 22 above)
Comment (glwagner) [reply test/test_reactant_correctness.jl]: unresolve if not true!

**Comment by glwagner** [comment]
The regression tests are failing bc of a prior issue (https://github.com/CliMA/Oceananigans.jl/issues/5354) so I will merge this!
